BrightID integration #2

pettinarip opened the pull request:
Iteration #2 for BrightID integration.

BrightID testing tool https://acolytec3.github.io/brightid_test_app/
- Context: CLRFundTest
- Test api key: 55HhFtQvaHB0BJeR

pettinarip committed:
wire verify ui with bright id store state

pettinarip committed:
check unique flag on brightid verification return payload

pettinarip committed:
first connection between brightid state and UI

pettinarip commented on vue-app/src/api/bright-id.ts:
```diff
@@ -82,3 +87,36 @@ export async function registerUser(
+    const verification = await getVerification(contextId)
+    brightId.isLinked = true
+    brightId.isSponsored = true
+    brightId.isVerified = !!verification?.unique
```
@corwintines this is one of the main changes you need to be worried about if you branch your work off `develop`. I would suggest to start from this branch and merge everything here.

samajammin commented on vue-app/.env.example:
```diff
@@ -11,6 +11,8 @@ VUE_APP_CLRFUND_FACTORY_ADDRESS=0x5FC8d32690cc91D4c39d9d3abcBD16989F875707
 # Supported values: simple, brightid
 VUE_APP_USER_REGISTRY_TYPE=simple
+# clr.fund (prod) or CLRFundTest (testing)
+VUE_APP_BRIGHTID_CONTEXT=clr.fund
```
Good call! We should also use this in our deploy scripts when we pass context into the registry constructor, e.g. https://github.com/ethereum/clrfund/blob/develop/contracts/scripts/deploy.ts#L45

samajammin replied:
Not sure if we have to encode the string in any way?

pettinarip replied:
Yes, I was just starting to look into that. I think we need to encode it, you are right.

wackerow committed:
layout adjustments for verify-landing

Removes background scroll on larger screens. Keeps image fixed at bottom on smaller screens with scroll room to allow content visibility.

wackerow committed:
fix logic for isRoundFullOrOver on verify-landing

Adds getter checking if message cap has been hit, which would block user from contributing. Removes check for isRecipientRegistryFull, as this wouldn't stop a user from signing up and contributing if the message cap hasn't been hit.

wackerow committed:
updates verify-landing copy with note about two txs

wackerow committed:
changes "go home" to "go back"

In line with existing design, link returns back to projects page

wackerow committed:
adds tooltip with copy about reusing BrightID account

corwintines committed:
feat: BrightIdWidget getters and use cases

pettinarip committed:
adjust sponsor and register steps

pettinarip committed:
add brightid parties link

pettinarip committed:
cleanup code

pettinarip committed:
check if the user is verify button

pettinarip committed:
remove clickable styles on nav links

pettinarip committed:
move brightid loading logic out to its own action

pettinarip committed:
cleanups and mobile fixes

pettinarip committed:
remove tooltips

pettinarip committed:
read brightid context from env

pettinarip commented:
Sorry guys for the big PR :S

Here are a couple of things that we still need to polish:
- Get Verified step
	- [x] Check copies. Is the "Check if you are verified" button text ok?
- Get Registered step
	- [x] ~~Figure out what the `verifier` address should be in the `BrightIdUserRegistry` contract constructor. Currently, I'm receiving a `'NOT AUTHORIZED'` error when doing the tx.~~
- [ ] URL paths: currently we have the `/verify/connect` path by default. As we are not going to synchronize the URL with the inner steps anymore, should we change the path to something like `/verifying`? or maybe handle everything inside the `/verify` route...where we display the landing page by default.

pettinarip committed:
fix vue warning about mutating props

pettinarip committed:
set brightid verifier address

pettinarip commented on contracts/.env.example:
```diff
@@ -3,6 +3,10 @@ RECIPIENT_REGISTRY_TYPE=simple
+# clr.fund (prod) or CLRFundTest (testing)
+BRIGHTID_CONTEXT=clr.fund
+# BrightId node addr that signs verifications. Node One uses this one
+BRIGHTID_VERIFIER_ADDR=0xb1d71F62bEe34E9Fc349234C201090c33BCdF6DB
```
More info can be found in the BrightID smart contract sample README, here: https://github.com/BrightID/BrightID-SmartContract

Shared doc with good technical info: https://docs.google.com/document/d/16-gqjfJLz7WSYVDHVQSFKK3Mj0GFWetxdgE9nQuMS9o

pettinarip commented on vue-app/src/store/index.ts:
```diff
@@ -287,17 +287,6 @@ const actions = {
         state.currentUser.walletAddress
       )
 
-      // If the user is registered, we assume all brightId steps as done
```
@samajammin I've separated all this brightId fetching logic into its own action (see below, LOAD_BRIGHT_ID). Why? because:
1. We have a polling made with the LOAD_USER_INFO action in the `App.vue` level and I felt that it was unnecessary to fetch brightid data all the time (unless you require to do it as in the user verification UI).
2. In the verification UI, we need to do a polling on the brightId data.

samajammin replied:
Nice, ya I was also thinking this is probably worth breaking out separately.

pettinarip committed:
load a default state of brightid on init

wackerow committed:
checks for max contributors reached

samajammin commented on vue-app/src/api/bright-id.ts:
```diff
@@ -6,7 +6,21 @@ import { BrightIdUserRegistry } from './abi'
+export interface BrightId {
+  isLinked: boolean
+  isSponsored: boolean
+  isVerified: boolean // If is verified in BrightID
```
```suggestion
  isVerified: boolean
```

samajammin commented on vue-app/src/api/bright-id.ts:
```diff
@@ -30,17 +44,10 @@ export async function selfSponsor(
 export function getBrightIdLink(userAddress: string): string {
   const nodeUrl = 'http:%2f%2fnode.brightid.org'
-  const deepLink = `brightid://link-verification/${nodeUrl}/${CONTEXT}/${userAddress}`
+  const deepLink = `brightid://link-verification/${nodeUrl}/${process.env.VUE_APP_BRIGHTID_CONTEXT}/${userAddress}`
```
Should we default to 'clr.fund' here?

samajammin commented on vue-app/src/api/bright-id.ts:
```diff
@@ -55,14 +62,14 @@ export class BrightIdError extends Error {
-): Promise<Verification | null> {
-  const apiUrl = `${NODE_URL}/verifications/clr.fund/${userAddress}?signed=eth&timestamp=seconds`
+): Promise<Verification> {
+  const apiUrl = `${NODE_URL}/verifications/${process.env.VUE_APP_BRIGHTID_CONTEXT}/${userAddress}?signed=eth&timestamp=seconds`
```
Should we default to 'clr.fund' here?

samajammin commented on vue-app/src/api/bright-id.ts:
```diff
@@ -73,7 +80,7 @@ export async function registerUser(
   const registry = new Contract(registryAddress, BrightIdUserRegistry, signer)
   const transaction = await registry.register(
-    formatBytes32String(CONTEXT),
+    formatBytes32String(process.env.VUE_APP_BRIGHTID_CONTEXT || ''),
```
Should we default to 'clr.fund' here?

samajammin commented on vue-app/src/store/index.ts:
```diff
@@ -326,14 +315,40 @@ const actions = {
       commit(SET_CURRENT_USER, {
+        // By default, when a currentUser has not been loaded before, set all
+        // brightid states to false
+        brightId: {
```
But won't this now also overwrite the BrightID state every time `LOAD_USER_INFO` is dispatched (which we trigger every 60 seconds in the App.vue)?

samajammin replied:
I feel like we should just set this to whatever the current state values are.

pettinarip replied:
The default `brightId` object then gets overwritten by the spread of `...state.currentUser` (if there is something to spread).

BUT yes...when I did it, I didn't like it either. Honestly, I was just doing it to avoid the optional `brightId?: BrightId` in the interface definition of the User.

I think we could leave it as `brightId?` and clean this ugly thing I did.

samajammin commented on vue-app/src/components/WalletModal.vue:
```diff
@@ -92,6 +93,7 @@ export default class WalletModal extends Vue {
     if (this.$store.state.currentRound) {
       // Load cart & contributor data for current round
       this.$store.dispatch(LOAD_USER_INFO)
+      this.$store.dispatch(LOAD_BRIGHT_ID)
```
I wonder - should we add an `await` to `this.$store.dispatch(LOAD_USER_INFO)`?

Looks like the `LOAD_BRIGHT_ID` action depends on having `state.currentUser.walletAddress` in order to fetch the  data.

pettinarip replied:
Good catch! yes, I'll change that.

samajammin committed:
Add default BrightID context

samajammin committed:
Add default BRIGHTID_VERIFIER_ADDR

samajammin committed:
Add new ENV variables to deployRound script

samajammin commented on vue-app/src/views/Verify.vue:
```diff
@@ -431,188 +312,165 @@ export default class VerifyView extends mixins(validationMixin) {
+  /**
+   * Start polling brightId state until the condition is met
+   */
+  private async waitUntil(isConditionMetFn, intervalTime = 5000) {
```
😍

samajammin reviewed (changes requested):
A couple thoughts... LMK what you think!

samajammin commented:
> Is the "Check if you are verified" button text ok?

@pettinarip button looks good but I think we should give some feedback (i.e. spinner while fetching, some text info if we fetched the data & the user is still not verified). Currently if user clicks this button, nothing happens in the UI if they're still not verified.

(Happy for us to solve this in a separate PR)

![Image 2021-08-09 at 5 04 39 PM](https://user-images.githubusercontent.com/8097623/128789949-c4b89831-e972-4b83-a5bb-47ed7fd01682.jpg)

samajammin committed:
Fix isRoundFullOrOver logic

samajammin committed:
Merge pull request #232 from ethereum/brightid/landing

BrightID landing page cleanup

samajammin committed:
Fix CallToActionCardGetters

pettinarip commented:
> @pettinarip button looks good but I think we should give some feedback (i.e. spinner while fetching, some text info if we fetched the data & the user is still not verified). Currently if user clicks this button, nothing happens in the UI if they're still not verified.

Absolutely! I thought that we could receive the verification and display the success verification message for 2 or 3 seconds before we transition to the next step (mainly because Ryan left some UI for this particular case and could leverage from that).

pettinarip committed:
check for brightid verification and registration to display proper messages to the user

pettinarip committed:
await user info and brightid loads before loading the cart state

pettinarip committed:
remove default brightid state and mark it as optional in the User state

pettinarip committed:
Merge branch 'brightid-integration-2' of github.com:ethereum/clrfund into brightid-integration-2

pettinarip committed:
set loading state and success verification UI when the user gets verified

samajammin committed:
Merge pull request #246 from ethereum/brightid-cart

Cart - Check BrightId state

samajammin committed:
Fix isBrightIdRequired

samajammin committed:
Fix isRoundFullOrOver

samajammin reviewed (approved):
Looks great! Pushed up a couple small changes.

Let's just wait on https://github.com/ethereum/clrfund/pull/233 before merging this.

corwintines committed:
Merge branch 'brightid-integration-2' into brightid-widget

pettinarip committed:
add brightid readme

corwintines committed:
fix: change requests

corwintines committed:
fix: add click functionality where missed for closing profile'

samajammin committed:
Clean up widget styles within profile

samajammin committed:
Fix fixed icon

samajammin committed:
Update CallToActionCard logic for BrightID

samajammin committed:
Simplify Cart BrighID logic

samajammin committed:
Merge pull request #233 from ethereum/brightid-widget

BrightIdWidget

samajammin committed:
Add BrightID resource links

samajammin reviewed (approved):
Bravo 👏 👏 👏
